### PR TITLE
WIP: Fix/convert attention

### PR DESCRIPTION
**Pull request by wenshuoliu**
- add 2-head attention concatenation
- add layernorm and skip connection to ffn
- fix reduction mask

### CODE REVIEW

**Comment by dpressel** [comment]
going to leave this as a WIP for now and merge later assuming it fixes the issues we are seeing

**Comment by wenshuoliu** [comment]
Made a couple of changes: 
- GeLU was not fed in before
- the output encoding is to be fed into cosine distance. I don't think it make sense to let them go through an activation.

File: api-examples/pretrain-paired.py [hunk 1]
```diff
@@ -72,11 +74,14 @@ def train():
                         help="Path or url of the dataset cache")
     parser.add_argument("--d_model", type=int, default=512, help="Model dimension (and embedding dsz)")
     parser.add_argument("--d_ff", type=int, default=2048, help="FFN dimension")
-    parser.add_argument("--d_k", type=int, default=None, help="Dimension per head.  Use if num_heads=1 to reduce dims")
+    parser.add_argument("--d_k", type=int, default=64, help="Dimension per head.  Use if num_heads=1 to reduce dims")
```
Comment (wenshuoliu) [inline api-examples/pretrain-paired.py:77]: add a new option to reduction layer, and document it.

File: api-examples/response-selection.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
+import argparse
```
Comment (wenshuoliu) [inline api-examples/response-selection.py:1]: this go in transformer.utils

File: api-examples/response-selection.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+from transformer_utils import *
+from baseline.pytorch.lm import TransformerLanguageModel
+from baseline.progress import create_progress_bar
+from baseline.utils import str2bool
+from torch.utils.data import DataLoader
+import logging
+logger = logging.getLogger("baseline")
+
+
+def create_model(model_type, embeddings, d_model=512, d_ff=2048, dropout=0., num_heads=8, num_layers=6,
```
Comment (wenshuoliu) [inline api-examples/response-selection.py:12]: create_paired

File: api-examples/response-selection.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                                                layers=num_layers,
+                                                rpr_k=rpr_k,
+                                                d_k=d_k,
+                                                src_keys=['x'],
+                                                tgt_key='x')
+    # logger.info(model)
+    return model
+
+
+def get_joint_logprob(model: nn.Module, sequences: torch.Tensor, context_length: int):
```
Comment (wenshuoliu) [inline api-examples/response-selection.py:50]: move to transformer.utils

File: api-examples/transformer_utils.py [hunk 5]
```diff
@@ -92,6 +92,190 @@ def forward(self, inputs, targets):
...
+
+        :return: The final layer
+        """
+        x = self.ln1(inputs)
+        for layer in self.layer_stack:
+            x = layer(x)
+        x = self.final(x)
+        if self.skip_connect:
+            x = x + self.proj(inputs)
+        return self.ln2(x)
```
Comment (wenshuoliu) [inline api-examples/transformer_utils.py:191]: ln2 shouldn't be there

File: api-examples/transformer_utils.py [hunk 6]
```diff
@@ -92,6 +92,190 @@ def forward(self, inputs, targets):
...
+
+
+class SingleHeadReduction(nn.Module):
```
Comment (wenshuoliu) [inline api-examples/transformer_utils.py:194]: put it in the eight_mile.pytorch.layers

File: api-examples/transformer_utils.py [hunk 7]
```diff
@@ -92,6 +92,190 @@ def forward(self, inputs, targets):
...
+        self.attn = self.attn_fn.attn
+
+        x = x.squeeze(1)  # [B, T, D]
+        x = x * seq_mask.unsqueeze(-1)
+        x = x.sum(dim=1)  # [B, D]
+        x = x * seq_lengths.float().sqrt().unsqueeze(-1)
+        return x
+
+
+class TwoHeadConcat(nn.Module):
```
Comment (wenshuoliu) [inline api-examples/transformer_utils.py:254]: to eight_mile layers

File: api-examples/transformer_utils.py [hunk 8]
```diff
@@ -92,6 +92,190 @@ def forward(self, inputs, targets):
...
+
+class DenseLNStack(nn.Module):
+    """A stack of one or more hidden layers
+    """
+
+    def __init__(
+            self,
+            insz: int,
+            hsz: Union[int, List[int]],
+            activation: str = "gelu",
```
Comment (wenshuoliu) [inline api-examples/transformer_utils.py:143]: activation as a union of str and list

File: api-examples/transformer_utils.py [hunk 9]
```diff
@@ -92,6 +92,190 @@ def forward(self, inputs, targets):
...
+        * *hsz* -- (``int``) The number of hidden units (defaults to `100`)
+
+        :return: The final layer
+        """
+        x = self.ln1(inputs)
+        for layer in self.layer_stack:
+            x = layer(x)
+        x = self.final(x)
+        if self.skip_connect:
+            x = x + self.proj(inputs)
```
Comment (wenshuoliu) [inline api-examples/transformer_utils.py:190]: put it in residualblock. check in_feature of input.

Comment (wenshuoliu) [reply api-examples/transformer_utils.py:190]: util function to search in_features, in_channels, in_dims

Comment (wenshuoliu) [reply api-examples/transformer_utils.py:190]: not in the DenseStack, not general

File: api-examples/response-selection.py (hunk 2 above)
Comment (dpressel) [reply api-examples/response-selection.py:1]: ??? maybe wrong line number?

File: api-examples/transformer_utils.py (hunk 5 above)
Comment (dpressel) [reply api-examples/transformer_utils.py:191]: actually dont think 189-191 belongs in DenseStack.  Also DenseStack should be simplified to just call the sequential's forward

Comment (dpressel) [reply api-examples/transformer_utils.py:191]: you can get the same functionality in the dual encoder by having a `nn.Sequential(DenseStack, ResidualOutputLayer)`

File: api-examples/pretrain-paired.py (hunk 1 above)
Comment (dpressel) [reply api-examples/pretrain-paired.py:77]: I edont think we want d_k on by default

**Comment by dpressel** [comment]
closing as stale
